File: server/utils/queue/baseQueueManager.js
import logger from "../logger.js";

class BaseQueueManager {
    constructor(options = {}) {
        this.items = [];
        this.options = {
            readFunction: null,      // Function to read data from storage
            saveFunction: null,      // Function to save data to storage
            validateFunction: null,  // Function to validate items
            formatFunction: null,    // Function to format items
            duplicateCheckKey: null, // Key to check for duplicates
            ...options
        };
        this.initialize();
    }

    initialize() {
        if (this.options.readFunction) {
            this.items = this.readItems() || [];
        }
    }

    readItems() {
        try {
            return this.options.readFunction();
        } catch (error) {
            logger.error("Error reading items:", { error });
            return [];
        }
    }

    saveItems() {
        if (this.options.saveFunction) {
            try {
                this.options.saveFunction(this.items);
            } catch (error) {
                logger.error("Error saving items:", { error });
            }
        }
    }

    validateItem(item) {
        if (this.options.validateFunction) {
            return this.options.validateFunction(item);
        }
        return true;
    }

    formatItem(item) {
        if (this.options.formatFunction) {
            return this.options.formatFunction(item);
        }
        return item;
    }

    isDuplicate(item) {
        if (!this.options.duplicateCheckKey) return false;
        return this.items.some(existingItem => 
            existingItem[this.options.duplicateCheckKey] === item[this.options.duplicateCheckKey]
        );
    }

    add(item) {
        if (!this.validateItem(item)) {
            logger.error("Invalid item:", { item });
            return false;
        }

        if (this.isDuplicate(item)) {
            logger.warn("Duplicate item not added:", { item });
            return false;
        }

        const formattedItem = this.formatItem(item);
        this.items.push(formattedItem);
        this.saveItems();
        return true;
    }

    addToFront(item) {
        if (!this.validateItem(item)) {
            logger.error("Invalid item:", { item });
            return false;
        }

        if (this.isDuplicate(item)) {
            logger.warn("Duplicate item not added:", { item });
            return false;
        }

        const formattedItem = this.formatItem(item);
        this.items.unshift(formattedItem);
        this.saveItems();
        return true;
    }

    addMany(items, addToFront = false) {
        if (!Array.isArray(items)) {
            logger.error("Invalid input. Items must be an array.");
            return 0;
        }

        let addedCount = 0;
        const validItems = [];

        items.forEach(item => {
            if (this.validateItem(item) && !this.isDuplicate(item)) {
                const formattedItem = this.formatItem(item);
                validItems.push(formattedItem);
                addedCount++;
            }
        });

        if (addedCount > 0) {
            if (addToFront) {
                this.items.unshift(...validItems);
            } else {
                this.items.push(...validItems);
            }
            this.saveItems();
        }

        return addedCount;
    }

    removeFromFront() {
        if (this.items.length > 0) {
            const removedItem = this.items.shift();
            this.saveItems();
            return removedItem;
        }
        logger.warn("No items to remove from front");
        return null;
    }

    removeFromBack() {
        if (this.items.length > 0) {
            const removedItem = this.items.pop();
            this.saveItems();
            return removedItem;
        }
        logger.warn("No items to remove from back");
        return null;
    }

    removeAtIndex(index) {
        const actualIndex = index - 1; // Convert to 0-based index
        if (actualIndex >= 0 && actualIndex < this.items.length) {
            const removedItem = this.items.splice(actualIndex, 1)[0];
            this.saveItems();
            return removedItem;
        }
        logger.error("Invalid index or no items to remove");
        return null;
    }

    getFirst() {
        return this.items.length > 0 ? this.items[0] : null;
    }

    getLast() {
        return this.items.length > 0 ? this.items[this.items.length - 1] : null;
    }

    getAll() {
        return this.items;
    }

    clear() {
        this.items = [];
        this.saveItems();
    }

    getLength() {
        return this.items.length;
    }
}

export default BaseQueueManager;

File: app/managers/base_queue_manager.py
from typing import Callable, Optional, List, Any
from app.core import logger

class BaseQueueManager:
    def __init__(self, options: dict = None):
        if options is None:
            options = {}
        
        self.items = []
        self.options = {
            'read_function': None,
            'save_function': None,
            'validate_function': None,
            'format_function': None,
            'duplicate_check_key': None,
            **options
        }
        self.initialize()
    
    def initialize(self):
        if self.options['read_function']:
            self.items = self.read_items() or []
    
    def read_items(self):
        try:
            return self.options['read_function']()
        except Exception as error:
            logger.error("Error reading items", message=str(error))
            return []
    
    def save_items(self):
        if self.options['save_function']:
            try:
                self.options['save_function'](self.items)
            except Exception as error:
                logger.error("Error saving items", error=str(error))
    
    def validate_item(self, item):
        if self.options['validate_function']:
            return self.options['validate_function'](item)
        return True
    
    def format_item(self, item):
        if self.options['format_function']:
            return self.options['format_function'](item)
        return item
    
    def is_duplicate(self, item):
        if not self.options['duplicate_check_key']:
            return False
        key = self.options['duplicate_check_key']
        return any(
            existing_item.get(key) == item.get(key)
            for existing_item in self.items
        )
    
    def add(self, item):
        if not self.validate_item(item):
            logger.error("Invalid item", item=item)
            return False
        
        if self.is_duplicate(item):
            logger.warn("Duplicate item not added", item=item)
            return False
        
        formatted_item = self.format_item(item)
        self.items.append(formatted_item)
        self.save_items()
        return True
    
    def add_to_front(self, item):
        if not self.validate_item(item):
            logger.error("Invalid item", item=item)
            return False
        
        if self.is_duplicate(item):
            logger.warn("Duplicate item not added", item=item)
            return False
        
        formatted_item = self.format_item(item)
        self.items.insert(0, formatted_item)
        self.save_items()
        return True
    
    def add_many(self, items, add_to_front=False):
        if not isinstance(items, list):
            logger.error("Invalid input. Items must be an array.")
            return 0
        
        added_count = 0
        valid_items = []
        
        for item in items:
            if self.validate_item(item) and not self.is_duplicate(item):
                formatted_item = self.format_item(item)
                valid_items.append(formatted_item)
                added_count += 1
        
        if added_count > 0:
            if add_to_front:
                self.items = valid_items + self.items
            else:
                self.items.extend(valid_items)
            self.save_items()
        
        return added_count
    
    def remove_from_front(self):
        if len(self.items) > 0:
            removed_item = self.items.pop(0)
            self.save_items()
            return removed_item
        logger.warn("No items to remove from front")
        return None
    
    def remove_from_back(self):
        if len(self.items) > 0:
            removed_item = self.items.pop()
            self.save_items()
            return removed_item
        logger.warn("No items to remove from back")
        return None
    
    def remove_at_index(self, index):
        actual_index = index - 1
        if 0 <= actual_index < len(self.items):
            removed_item = self.items.pop(actual_index)
            self.save_items()
            return removed_item
        logger.error("Invalid index or no items to remove")
        return None
    
    def get_first(self):
        return self.items[0] if len(self.items) > 0 else None
    
    def get_last(self):
        return self.items[-1] if len(self.items) > 0 else None
    
    def get_all(self):
        return self.items
    
    def clear(self):
        self.items = []
        self.save_items()
    
    def get_length(self):
        return len(self.items)
